src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs

```diff
@@ -254,6 +268,25 @@ public void CancelRun(bool force)
         {
             foreach (var shift in Shifts)
                 shift.Cancel(force);
+
+            if (force)
+            {
+                SpinWait.SpinUntil(() => _topLevelWorkItem.State == WorkItemState.Complete, WAIT_FOR_FORCED_TERMINATION);
+
+                // Notify termination of any remaining in-process suites
+                lock (_activeWorkItems)
+                {
+                    int index = _activeWorkItems.Count;
+
+                    while (index > 0)
+                    {
+                        var work = _activeWorkItems[--index];
+
+                        if (work.State == WorkItemState.Running)
+                            new CompositeWorkItem.OneTimeTearDownWorkItem(work).WorkItemCancelled();
+                    }
+                }
+            }
```

Comment:
This is the core of the fix. Dispatcher observes that the test suite threads didn't take care of completing themselves so it does it for them. IMO, active management by the dispatcher like this is really what we need, simplifying the code in the work items and centralizing the logic in the dispatcher. could be done incrementally or in one big bang by use of a feature switch.